### PR TITLE
Use sparse array in ALE, ALESubtraction, SCALE, KDA, and MKDADensity

### LINKED ISSUE #692: Handle sparse/dense uniformly for different kernels
https://github.com/neurostuff/NiMARE/blob/e0d7cf5dc939729d18c637f1d73ee694acdc6bac/nimare/meta/kernel.py#L202

This line is buggy, because only MKDA return a sparse matrix.

For ALE, this line returns a dense matrix.

In addition, it's worth thinking what we can do to speed things up, as going from data -> sparse then sparse -> costs time

### PR DESCRIPTION
**Pull request by JulioAPeraza**
Closes #692.

Changes proposed in this pull request:

- Create the MA maps for all studies in `compute_ale_ma` and return a 4D sparse array.
- Return sparse array by `ALEKernel._transform` and `KernelTransformer.transform`.
- Add support for `ma_maps` 4D sparse array in `ALE` and `CBMAEstimator`. This will involve adding support for sparse array in `ALE._compute_summarystat_est`, `ALE._determine_histogram_bins`, and `ALE._compute_null_approximate`.

### CODE REVIEW

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:243]: I'm not sure how well this generalizes to other masker types, but I've opened an issue requesting a new attribute for masker objects with this info: https://github.com/nilearn/nilearn/issues/3296

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:251]:
```suggestion
                # The first column of coords is the fourth dimension of the dense array
                study_ma_values = data[coords[0, :] == exp_idx]
```

File: nimare/meta/cbma/base.py
Comment (tsalo) [inline nimare/meta/cbma/base.py:160]: If we are enforcing a sparse array here, will `_compute_null_approximate`, `_compute_summarystat_est`, or `_determine_histogram_bins` ever actually encounter any of the other possible datatypes? If not, then I think there's some spaghetti code we can simplify. If yes, then I'd like to try to document the circumstances where that would happen.

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:317]: Thanks for including this! This is the only improvement from #709 that I wanted to get merged, so I'm glad I don't need to open a separate PR.

**Review by tsalo — COMMENTED** [review]
I noticed a couple of commented-out lines that will to be removed before this is merged. I had a few comments/suggestions as well. Thanks for this!

File: nimare/meta/utils.py
Comment (tsalo) [inline nimare/meta/utils.py:400]: I'd like to compare this to https://github.com/tsalo/NiMARE/blob/49191a0a4b10a2ac522828c9e558f279d5c6d6ea/nimare/meta/utils.py#L400-L438 before we merge.

File: nimare/meta/cbma/base.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:160]: Yeah, I think once we get all tests passing using the sparse array I will remove support for the other datatypes.

Comment (tsalo) [reply nimare/meta/cbma/base.py:160]: Is the PR ready for this, or are you planning to do more before removing the other datatypes?

Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:160]: I think it's almost ready. The [example figures](https://nimare--725.org.readthedocs.build/en/725/auto_examples/02_meta-analyses/01_plot_cbma.html#sphx-glr-auto-examples-02-meta-analyses-01-plot-cbma-py) in the documentation are similar to the ones in the [master](https://nimare.readthedocs.io/en/latest/auto_examples/02_meta-analyses/01_plot_cbma.html#sphx-glr-auto-examples-02-meta-analyses-01-plot-cbma-py) branch but are not the same. I need to look at the code to find what I'm missing, then rerun some profiling checks for ALE, KDA, and MKDADensity.

Comment (tsalo) [reply nimare/meta/cbma/base.py:160]: Ah, thanks for the clarification. In that case, I'll wait until you request my review before looking through everything.

Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:160]: I think I am ready to remove the support for other data types. I started doing that this morning but some tests were failing because they were created to work with numpy array. Should we remove those tests in that case?

I can also provide a table with the profiling results `ALE`, `KDA`, and `MKDADensity` against the master branch. At first glance, the use of sparse directly in the `CBMAEstimator` makes everything more efficient in terms of memory and computation time.
Also, `_run_fwe_permutation` efficiency got improved by using sparse.

Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:160]: @tsalo to remove return_type="array" in favor of return_type="sparse" I would need to implement sparse in ALESubtraction too. Do you think that can be done here or in another PR?

Comment (tsalo) [reply nimare/meta/cbma/base.py:160]: I think it's worth implementing it in this PR.

**Comment by JulioAPeraza** [comment]
@tsalo I believe this is ready for review. Now, all CBMA methods support sparse arrays.

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:165]:
```suggestion
            self.__n_mask_voxels = stat_values.shape[0]
```
I'm thinking this should either be private (leading underscores) or considered a fitted, data-dependent parameter (trailing underscore).

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:196]: Can you add a comment with info about _when_ this would be reached? I assume it would be reached when `ma_maps` is a list above, but if there are other cases then I'd like those documented as well.

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:651]: I think we're going to need to profile this.

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:504]: Should this be called after every `del`?

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:210]: Hm... should it be `if isinstance(...) and return_type != "sparse"`?

File: nimare/meta/utils.py
Comment (tsalo) [inline nimare/meta/utils.py:436]: I believe this is a dict now.

File: nimare/meta/cbma/ale.py
Comment (jdkent) [inline nimare/meta/cbma/ale.py:182]: could you add the possible types `ma_maps` can be in the docstring of `_determine_histogram_bins`? and maybe how many dimensions the array can have.

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py:159]: Here I try to follow the way the `KernelTransformer.transform()` creates the mask_data. Do you think we can easily add support for other makers here, or that could be addressed in a separate PR?

File: nimare/meta/cbma/mkda.py
Comment (jdkent) [inline nimare/meta/cbma/mkda.py:229]: This doc string may be out of date now, `ma_maps` can be sparse arrays as well.

File: nimare/meta/cbma/mkda.py
Comment (jdkent) [inline nimare/meta/cbma/mkda.py:1084]: this doc string should be updated too to include sparse arrays as input

File: nimare/meta/cbma/mkda.py
Comment (jdkent) [inline nimare/meta/cbma/mkda.py:1115]: could you update this docstring to include what datatypes `ma_maps` can be?

File: nimare/meta/cbma/mkda.py
Comment (jdkent) [inline nimare/meta/cbma/mkda.py:1174]: this doc string should be updated to include sparse arrays

File: nimare/meta/cbma/mkda.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/mkda.py:504]: For MKDAChi2 I had to call this because `del` wasn't freeing up the memory there. I will profile this line to see if it's needed here.

**Review by jdkent — CHANGES_REQUESTED** [review]
Looking great! I have a couple places where docstrings should likely be updated, and I'm testing locally to if some of the functions are ever called with `np.array` since the switch to sparse arrays.

File: nimare/meta/kernel.py
Comment (JulioAPeraza) [reply nimare/meta/kernel.py:210]: I think if `return_type == "sparse"` this line won't be reached, the transformer will return the transformed_maps in line 202:

https://github.com/JulioAPeraza/NiMARE/blob/bb9815891727cac66cbda5a80da60d7e0ad42c9b/nimare/meta/kernel.py#L202-L203

File: nimare/meta/utils.py
Comment (JulioAPeraza) [reply nimare/meta/utils.py:436]: We're creating the empty dictionary inside `compute_ale_ma` if `use_dict is True`. I think `kernels` will be a 3D array when `use_dict is False` and `sample_sizes is None`

File: nimare/meta/cbma/mkda.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/mkda.py:229]: I believe `ma_maps` aren't used by the `MKDADensity._compute_null_approximate()`. I think the `ma_maps` parameter is here just to provide compatibility when the `_compute_null_approximate()` is called in the `CBMAEstimator`. 
What would you recommend doing in this case?

Comment (JulioAPeraza) [reply nimare/meta/cbma/mkda.py:229]: How does this look?

```
Parameters
----------
ma_maps
    Modeled activation maps. Unused for this estimator.
```

File: nimare/meta/cbma/base.py
Comment (jdkent) [inline nimare/meta/cbma/base.py:216]: I think if this is a simple voxelwise niftimasker (not a labels masker or some sort of atlas masker), this should still return a sparse matrix, not a dense one. Inconsistent behavior to re-run the `fit` method twice and have the data types be different for each run.

**Comment by jdkent** [comment]
these lines no longer need to be in the code: https://github.com/neurostuff/NiMARE/blob/ac57801b288675e97b91d4356ab8aa83ced9edfc/nimare/tests/test_meta_ale.py#L48-L53

since we have removed memory limit

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py:661]: Do you think we should keep the memap array here? 
I profiled the computation time is faster and the memory usage is not that different.

File: nimare/meta/cbma/base.py
Comment (jdkent) [reply nimare/meta/cbma/base.py:223]: I'm looking into it, we can potentially extend/catch the niftimasker class to allow for a custom transformation, here is some draft code for the masking procedure (don't know if it currently prevents entire array from being loaded at once)
```
nonzero_idx = imgs.dataobj.nonzero()
all_data = imgs.dataobj[nonzero_idx]
all_coords = np.vstack(nonzero_idx)
sparse_data = sparse.COO(all_coords, all_data, shape=imgs.shape)
masked_sparse_data = sparse_data[mask_img_.dataobj.nonzero()].T
```

File: nimare/meta/cbma/mkda.py
Comment (jdkent) [reply nimare/meta/cbma/mkda.py:229]: okay, that sounds good to me for now

File: nimare/meta/cbma/base.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:223]: I ended up using the same approach we have in `compute_ale_ma` and in `compute_kda_ma`. Now loading the pre-generated MA maps is faster and memory efficient. 
Thanks!

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [reply nimare/meta/cbma/mkda.py:503]: Thanks!

File: nimare/meta/kernel.py
Comment (tsalo) [reply nimare/meta/kernel.py:210]: Ah, that makes sense.

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py:201]: I'm deleting these lines since `_collect_ma_maps` always returns a sparse array now.

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py]: I have added the check for the masker to `_preprocess_input` where the masker is called for the first time. In the initialization of the classes `dataset` doesn't exist yet, so I didn't know where to add the check.

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py]: Done!

File: nimare/meta/cbma/mkda.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/mkda.py:1180]: Done!

File: nimare/meta/cbma/ale.py
Comment (tsalo) [reply nimare/meta/cbma/ale.py]: I would recommend putting it in `_fit`, like we have for the IBMAEstimators:

https://github.com/neurostuff/NiMARE/blob/ac57801b288675e97b91d4356ab8aa83ced9edfc/nimare/meta/ibma.py#L169-L175

Comment (tsalo) [reply nimare/meta/cbma/ale.py]: But yes, I agree that it shouldn't be checked at initialization, since we don't have access to the Dataset yet.

File: nimare/meta/cbma/mkda.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/mkda.py:503]: I'm removing the `gc.collect()`, `del` is working fine here.

Comment (tsalo) [reply nimare/meta/cbma/mkda.py]: That is so confusing! I have no clue why it's necessary for some things, but not others... Thanks for checking it in this case though.

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py:647]:
```python
from nimare.dataset import Dataset
from nimare.meta.cbma.ale import SCALE
from nimare.utils import get_resource_path, vox2mm

dset_file = os.path.join(get_resource_path(), "nidm_pain_dset.json")
dset = Dataset(dset_file)

xyz = vox2mm(
    np.vstack(np.where(dset.masker.mask_img.get_fdata())).T,
    dset.masker.mask_img.affine,
)

meta = SCALE(xyz=xyz, n_iters=10)
results = meta.fit(dset)

```

This is what I got for 10 iterations:

- `master`:
    - 10/10 [00:03<00:00,  2.95it/s]
    - 228483/228483 [04:32<00:00, 839.13it/s]
    - Current memory usage is 18.818837MB; Peak was 231.680765MB
    - Computation time: 276.62519907951355

- `ale-sparse`:
    - 10/10 [00:06<00:00,  1.65it/s]
    - 228483/228483 [04:11<00:00, 908.33it/s]
    - Current memory usage is 29.558014MB; Peak was 198.31731MB
    - Computation time: 260.77495074272156

File: nimare/meta/cbma/ale.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/ale.py]: Done!

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:180]: I think this line can be removed, since it's the only possible type now.

File: nimare/meta/cbma/ale.py
Comment (tsalo) [inline nimare/meta/cbma/ale.py:211]: Can be removed as well.

File: nimare/meta/cbma/base.py
Comment (tsalo) [inline nimare/meta/cbma/base.py:158]: Can you cover this in one of the tests?

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:219]: This new null distribution needs to be documented in a few docstrings, I think.

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:1115]: Line can be removed.

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:1170]: Line can be removed.

File: nimare/meta/utils.py
Comment (tsalo) [inline nimare/meta/utils.py:432]:
```suggestion
        * This function now returns a 4D sparse array.
```

**Review by tsalo — CHANGES_REQUESTED** [review]
I think the core changes are ready. There are just a few docstring changes that I think need to be made, as well as a small improvement to one of the CBMA tests (i.e., covering the new masker type check).

File: nimare/meta/cbma/base.py
Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:158]: Do you know of any case in which this line is reached? If I provide a non-NiftiMasker to the estimator, it will get converted to a NiftiMasker by `get_masker(mask)`.

Comment (tsalo) [reply nimare/meta/cbma/base.py:158]: `get_masker` should preserve the type of any nilearn masker (e.g., `NiftiMasker`, `NiftiLabelsMasker`, `NiftiMapsMasker`), so you _should_ be able to feed in a NiftiLabelsMasker and it would then raise this exception.

Comment (JulioAPeraza) [reply nimare/meta/cbma/base.py:158]: Great. Thanks!

**Review by tsalo — APPROVED** [review]
Everything looks good to me, thanks!

**Pull request merged**
